mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp

```diff
@@ -160,6 +215,21 @@ struct LinalgOpTilingInterface
     return success();
   }
 
+  FailureOr<TilingResult> getTiledImplementationFromOperandTile(
+      Operation *op, OpBuilder &b, unsigned operandNumber,
+      ArrayRef<OpFoldResult> offsets, ArrayRef<OpFoldResult> sizes) const {
+    SmallVector<OpFoldResult> mappedOffsets, mappedSizes;
+    auto tilingInterfaceOp = cast<TilingInterface>(op);
+    if (failed(tilingInterfaceOp.getIterationDomainTileFromOperandTile(
+            b, operandNumber, offsets, sizes, mappedOffsets, mappedSizes))) {
+      return emitError(
+          op->getLoc(),
+          "unable to obtain the iter domain position of the operation.");
+    }
+    return tilingInterfaceOp.getTiledImplementation(b, mappedOffsets,
+                                                    mappedSizes);
```

Comment:
I think the cause of the problem I noted here: https://github.com/llvm/llvm-project/pull/88712/files#r1605830351

is because here this is returning a tiled implementation based on offsets and sizes of the iteration space, but does not know that one of the operands is already tiled. Concretely, we start with something like this
```
%2 = scf.for iter_args(%dest) {
  %0 = some_producer
  %1 = tensor.insert_slice %0 into %dest
  scf.yield %1
}
%3 = some_tilable_op %2
```
It gets (intermediately) rewritten to something like this
```
%3 = scf.for iter_args(%dest) {
  %0 = some_producer
  %1 = some_tilable_op %0 # INVALID, not yet tiled
  %2 = tensor.insert_slice %1 into %dest
  scf.yield %2
}
```
And then the tiled implementation is generated
```
%3 = scf.for iter_args(%dest) {
  %0 = some_producer
  %slice = tensor.extract_slice %0
  %1 = some_tiling_result %slice
  %2 = tensor.insert_slice %1 into %dest
  scf.yield %2
}
```

This API looks to be enforcing a slightly different contract than `getTiledImplementation` requires, namely that the tiled implementation must produce the requested tile AND consume the requested operand tile. I don't have a great immediate suggestion that doesn't involve rewriting `getTiledImplementation`. Maybe we can add an optional operand map to `getTiledImplementation` that maps from operand index to operand tile?

Reply:
```
%2 = scf.for iter_args(%dest) {
  %0 = some_producer
  %1 = tensor.insert_slice %0 into %dest
  scf.yield %1
}
%3 = some_tilable_op %2
```

From my understanding, the result through current implementation will look like:

```
%2, %3 = scf.for iter_args(%dest1, %dest2) {
  %0 = some_producer
  %1 = tensor.insert_slice %0 into %dest1
  %4 = tensor.extract_slice %1
  %5 = tiled_consumer %4
  %6 = tensor.insert %5 into %dest2
  scf.yield %1, %6
}
```

> Maybe we can add an optional operand map to getTiledImplementation that maps from operand index to operand tile?

Besides this method, I think another possible option is that we can directly RESET corresponding OPERAND of tiled consumer from the result of `extractSliceOp`_(generated by new tiled consumer)_ to the SOURCE of `InsertSliceOp`_(generated by tiled producer)_ in function `replaceInsertSliceWithTiledConsumer` where all necessary information is avaliable and it just takes one line code change. Then, the resulting IR will become:

```
%2, %3 = scf.for iter_args(%dest1, %dest2) {
  %0 = some_producer
  %1 = tensor.insert_slice %0 into %dest1
  %4 = tensor.extract_slice %1 // expect to be simplified later
  %5 = tiled_consumer %0 // reset `%4` to `%0`, which is the source of `%1` and actual tiled result of producer,
  %6 = tensor.insert %5 into %dest2
  scf.yield %1, %6
}
```

Reply:
Thanks for the catch @qedawkins !

I've addressed it in the way similar to what @Yun-Fly has suggested as well. Please take a look at the IR now.

Reply:
> Besides this method, I think another possible option is that we can directly RESET corresponding OPERAND of tiled consumer from the result of extractSliceOp(generated by new tiled consumer) to the SOURCE of InsertSliceOp

This solution appears problematic because `getTiledImplementation` is only defined in terms of a tile of the iteration space and the tile of the result that it returns. It makes no requirement that the tiling implementation uses `tensor.extract_slice` to extract the slice of the tensor operand that it wants.

Moreover this IR is UB unless %0 is forwardable to %4 as it is potentially reading uninitialized data.
```
%2, %3 = scf.for iter_args(%dest1, %dest2) {
  %0 = some_producer
  %1 = tensor.insert_slice %0 into %dest1
  %4 = tensor.extract_slice %1
  %5 = tiled_consumer %4
  %6 = tensor.insert %5 into %dest2
  scf.yield %1, %6
}
```
So the only way that I see to guarantee valid tiling is if `getTiledImplementationFromOperandTile` takes the operand tile it is expected to consume as an input.

Reply:
> Moreover this IR is UB unless %0 is forwardable to %4 as it is potentially reading uninitialized data.
> %2, %3 = scf.for iter_args(%dest1, %dest2) {
  %0 = some_producer
  %1 = tensor.insert_slice %0 into %dest1
  %4 = tensor.extract_slice %1
  %5 = tiled_consumer %4
  %6 = tensor.insert %5 into %dest2
  scf.yield %1, %6
}

As you said above, the root cause comes from using `tensor.extract_slice` to extract the tiled result. But NOTE that, this IR is not the FINAL result, after redirecting operand of tiled consumer, there is no use for `%4` any more as I illustrated [above](https://github.com/llvm/llvm-project/pull/88712#:~:text=Then%2C%20the%20resulting%20IR%20will%20become%3A). 

> Maybe we can add an optional operand map to getTiledImplementation that maps from operand index to operand tile?

I guess what you mean here is also using operand tile instead of `extract_slice` inside `getTiledImplementation`. So the difference looks like where we replace it? If wrong, please correct me. BTW, If invalid temporary IR is already not acceptable, I think it began since cloning consumer after `candidateSliceOp` and reset its operand [here](https://github.com/llvm/llvm-project/blob/41123a2871c752cdbbdece54cd7fb11a9a95c768/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp#L1407).

> So the only way that I see to guarantee valid tiling is if getTiledImplementationFromOperandTile takes the operand tile it is expected to consume as an input.

I know your point. However, compared with one line fix, it may take more time or changes involving more than one function I guess. 

Also, except for this issue you had pointed, please assume another complex use case:

```
%0 = linalg.op -> tensor<256,256>
%1 = scf.forall(%dest1) {
    extract_slice
    %2 = scf.for iter_args(%dest2){
           extract_slice    
           %3 =  tiled_producer  
           %4 = insert_slice %3 into %dest2 [of11, of12][sz11, sz12][1,1] tensor<32,32> to tensor<128,128>
           yield %4
    }
   scf.forall.in_parallel {
      parallel_insert_slice %2 into %dest1 [of21, of22][sz21, sz22][1,1] tensor<128,128> to tensor<256,256>
   }
}
// consumer to be fused
%5 = linalg.add(%0, %1: tensor<256,256>, tensor<256,256>) -> tensor<256,256>
```
Given current implementation, even if we correctly address one of the operand tile(like `%1`) of `linalg.add` like you said, it is hard to deal with another one(like `%0`) because the argument `offset` and `size` of `getTiledImplementationFromOperandTile` are both calculated by `candidateSliceOp`(`%4 = insert_slice %3`) based on intermediate buffer(`tensor<128,128>`), not coordinated on the original one(`tensor<256,256>`). Then, they will be used to infer other operand tile info under current implementation, which consequently leads to wrong or unmatched `offset` and `size` for `%0`, similar to the issue you found.

An optional way is that we can calculate the real `offset` and `size` coordinated on the original one without any tile in `replaceInsertSliceWithTiledConsumer`(need enhanced) before calling `getTiledImplementationFromOperandTile`, and that is also why I suggest to reset corresponding operand there.

Although, this scenario maybe not occur due to current limitation of implementation, it is a issue to be solved especially when we want to extend this feature to nested loop cases.

Anyway, it is up to author and you how to address issue you raised in this PR.

Reply:
Hi @qedawkins

So to deal with the above situation better, I've created two APIs now - `getTiledImplementationAsProducer` and `getTiledImplementationAsConsumer` - this helps us get around the issue you pointed out earlier. Please take a look.

Reasons for doing this :-
1. Previously having just `getTiledImplementation` worked because we didn't have use case for fusing as a consumer, but now that we do, this distinction in the function naming sounds a better design to me. Also, downstream integration can simply replace every `getTiledImplementation` with `getTiledImplementationAsProducer`.
2. During the downstream integration we don't want to add additional overhead of sending `std::nullopt` in all invocations of `getTiledImplementation` - please let me know if there's a way that I can achieve the following if you feel the above design is not welcomed :-
```
API: getTiledImplementation(Operation*, Builder, Offsets, Sizes, Optional<DenseMap<uint64_t, Value>>)

User:
// As a producer.
getTiledImplementation(op, b, offset, size);
// As a consumer.
getTiledImplementation(op, b, offset, size, operandTileMap);
```

I tried using `::std::optional<::llvm::DenseMap<uint64_t, Value>>` in [TilingInterface.td](https://github.com/llvm/llvm-project/blob/375761bcaba9ba5694890ece6eed5db286fc4fd1/mlir/include/mlir/Interfaces/TilingInterface.td#L71) and then tried assigning a default value (`std::nullopt`) to the argument when defining the function definition [here in TilingInterfaceImpl.cpp](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp#L115-L117), it required every invocation to explicitly send in that argument during invocation. Perhaps I'm missing some ODS specific declaration instead where this default value (`std::nullopt`) needs to be assigned ? Will need pointers for this in case the current design is not welcomed.

Reply:
> As you said above, the root cause comes from using tensor.extract_slice to extract the tiled result. But NOTE that, this IR is not the FINAL result, after redirecting operand of tiled consumer, there is no use for %4 any more as I illustrated https://github.com/llvm/llvm-project/pull/88712#:~:text=Then%2C%20the%20resulting%20IR%20will%20become%3A. You can also check the latest test mlir [result](https://github.com/llvm/llvm-project/blob/41123a2871c752cdbbdece54cd7fb11a9a95c768/mlir/test/Interfaces/TilingInterface/tile-and-fuse-consumer.mlir#L102-L110) whether meet your expectation?

The problem is not that there was temporarily invalid IR (that's common to have in patterns that manipulate regions) but instead that the consumer fusion pattern was responsible for fixing the potentially invalid IR in a way that was dependent on the tiling implementation.

> During the downstream integration we don't want to add additional overhead of sending std::nullopt in all invocations of getTiledImplementation - please let me know if there's a way that I can achieve the following if you feel the above design is not welcomed :-

Agree. I'm definitely open to other suggestions but what you added here makes the most sense to me. The other option is to omit the default implementation of `getTiledImplementationFromOperandTile` and do the extract_slice fixup per tiling implementation, but for that I'd like feedback from @ftynse and @MaheshRavishankar following the discussion above (although I don't think that this new API is even a valid composition of `getTiledImplementation` and `getIterationDomainTileFromOperandTile`).

Reply:
In general, I am not very happy to introduce more methods to an interface, especially if they are quite similar but subtly different. This may lead to copy-pasted implementations, divergent implementations and other things that are better avoided. The way I think about interfaces is a contract that another person will have to implement to opt into the transformation. The more complex the contract is, the less beneficial it is for that person to use it instead of just duplicating the transformation, which defies the purpose of sharing the transformation logic through interfaces.

I asked a question elsewhere in the patch driven by similar logic: why isn't `getTiledImplementationFromOperandTile` is a separate interface method and not a simple composition of `getIterationDomainTileFromOperandTile` and `getTiledImplementation` (from the iteration domain tile). IMO, we should orthogonalize the interface better. The tilable ops are fundamentally an iteration space + mappings from that space to indexing spaces of operands and results. In case of tilling, if we want to reason about the whole computation, the iteration space is a product of the loop iteration space and the iteration space of the tiled op.

Now, IIUC the problem is that `getTiledImplementation` currently assumes operating on the complete, untiled op which accepts full operands, and uses that to produce tiles of the operands. This assumption doesn't hold in the case when the consumer is forcibly moved in the loop. We seem to all agree that this assumptions should be lifted.

My thinking here is that `getTiledImplementation` does too much. It should not be emitting IR that computes a slice of each operand. Instead, it should be _given_ such slices. These slices in turn can be obtained from a separate `getOperandTileForIterationDomainTile` method (though I'm not sure whether this should be a method on the operation interface, or we should introduce a "TileableData" type interface that is implemented by tensors and potentially other types). This method would be called by "pure" tiling and producer fusion, and would not be called by consumer fusion _for the operand that comes from the producer_ (it will still be called for _other_ operands). Instead consumer fusion will directly supply `getTiledImplementation` with the slice of the operand that is just the result of the tiled producer (before `insert_slice`). This would also work if we were to introduce, e.g., nested linalg generics that folks have been discussing for years, where the data tiling of operands would be part of the surrounding generic that acts as a loop; as well as for the cases of non-hyperrectangular data structures.

This generally goes in a similar direction that what's proposed above, so I hope we are all converging. This will inflict some pain on downstreams, but looks like the more future-proof direction for upstream, and a better orthogonalization. (It is also unclear to me that having to implement similar-but-different methods for downstream other than yours is a significantly lesser pain than adapting to this.)  Specifically, we are decoupling the tiling of the _iteration space_ (operation) from the that of _data spaces_ (operands), and define the interface in terms that are related to tiling itself (iteration space tile, operand tile, result tile) rather than fusion (implementation as consumer, as producer), it's a tiling interface after all!

My preferred way to move this forward, if we all agree on the proposed decomposition of methods, is to make a separate PR that splits out `getOperandTileForIterationDomainTile` and uses it in the existing case, and rebase this PR on top of that. Though I could also accept these changes directly in this PR if rebasing is too involved.

There are also some observations for future generalizations that are non-blocking but appeared to me as I was thinking about this problem:
 -  What fusion needs to achieve is to "align" the iteration spaces of the producer and the consumer tilable ops. How do we check if this iteration space tile can be obtained from the original iteration space of the consumer op by tiling, e.g., if the consumer shifts the operand indexing as `i+1` or otherwise reuses it, this may not be trivially valid to just move it into the loop. Currently, we are likely restricting this to trivial permutations...
 - How (and where) do we generalize tiling of the result rather than always doing `insert_slice`?

Reply:
Ok, I think this PR has been put through the review wringer a bit too much. I dont think we should use PRs as a way to bike shed ideas. Maybe some clarification of what the interface is set up to do might help restrict the scope of expansions/modifications to first land this, and then think of how to change the interface to handle it. In general I have strong concerns with the direction of changes suggested in this comment thread. Ill answer those later, but first let me write down what each method in the interface does.

1) `getTiledImplementation` is designed to be as less opinionated as possible. It deals with just tiling of an operation. The operation publishes an iteration space, and given an tile of that iteration space, this method is meant to generate the tiled implementation of that operation. In process it might take slices of the operation/it might not, that is left to the implementation of the operation. This is the most basic method an operation has to implement for it to be tilable, and is potentially the most general way you can achieve tiling. The tiling algorithm is then very simple 
   a) Use the iteration domain to produce the inter-tile iteration space
   b) Then in the inner most loop body, where you know the tile of the iteration space, pass that tile information to this method to generate the loop body.

2) `getResultTilePosition` is also needed for implementing tiling. For a given tile of the iteration space, return the tile of the result that is produced by the tiled implementation. Most tensor tiling algorithms need to stitch back the original tensor value from the tiled operation. This method gives you back information that you need to stitch back the original tensor from the tiled implentation (by inserting the tiled result in the right place of the result tensor)

This just takes care of Tiling.  The following methods take care of fusion

1) `generateResultTileValue` deals with tile and fuse, specifically fusion of tiled consumer with untiled producer. Again it tries to be as less opinionated as possible about the operation semantics. All  this method expects is, given a tile of the result to be computed, generate a tiled implementation of the producer that computes the tile used by the consumer. If the operation cannot do it, it returns a failure. This method might or might not reuse the implementation in `getTiledImplementation`. That depends on op semantics. That is why there isnt a default implementation of this method in the current interface definition.

This is all you need for tile + fuse of producer. Now when we come to tile and fuse of producer, the flip of this is all we should need. IIRC this is what this patch initially started with but through many many rounds of reviews has deflected from initial intent (which is now lost in all the comment threads). All we should need for fusion of the tiled producer with untiled consumer is the following

1) `generateOperandTileValue`, i.e. given a tile of an operand, generate the implementation of the consumer that uses exactly this tile of the operand (I confess the name here is not great, open to naming suggestions). This method again might or might not reuse the implementation in `getTiledImplementation`. That again depends on op semantics which is a black box. If it cant do so, it returns a failure, and the fusion cannot happen. 

I think things get muddled when we try to have a default implementation of the two fusion methods. Default implementation now need to take an opinion on how the tiling of each operation works. I am fine with having a default implementation for the above methods, but IMO they are not core to the interface. They are just convenience methods that reduce implementation burden on operations that implement the interface. If it doesnt fit the semantics of the operation any operation can override this. Some helper methods that could be added

1) `getIterationSpaceTileFromResultTile` - given information of a tile of the result, compute the tile of the iteration space that when used with `getTiledImplementation` generates the required result tile. This can then be used to have a default implementation of `generateResultTileValue`

2) `getIterationSpaceTileFromOperandTile` - given information of a tile of the operand, compute the tile of the iteration space that when used with `getTiledImplementation` generates the tiled implementation that corresponds to the use of exactly the tile of the operand. This can be used to have a default implementation of `generateOperandTileValue`.

I dont think we need to change the `TilingInterface` fundamentally. All the base methods are meant to be as unopinionated as possible so as to allow more operations to be tiled (and fused based on methods the operation implements).

I suggest we first try to land this PR with something that is equivalent of `generateOperandTileValue` and then come back to maybe having default implementation of the different methods.

Reply:
> My thinking here is that `getTiledImplementation` does too much. It should not be emitting IR that computes a slice of each operand. Instead, it should be _given_ such slices. These slices in turn can be obtained from a separate `getOperandTileForIterationDomainTile` method (though I'm not sure whether this should be a method on the operation interface, or we should introduce a "TileableData" type interface that is implemented by tensors and potentially other types). This method would be called by "pure" tiling and producer fusion, and would not be called by consumer fusion _for the operand that comes from the producer_ (it will still be called for _other_ operands). Instead consumer fusion will directly supply `getTiledImplementation` with the slice of the operand that is just the result of the tiled producer (before `insert_slice`). This would also work if we were to introduce, e.g., nested linalg generics that folks have been discussing for years, where the data tiling of operands would be part of the surrounding generic that acts as a loop; as well as for the cases of non-hyperrectangular data structures.

I have strong concerns with this. `getTiledImplementation` is necessarily black boxed. I would not want to assume that you can produce an implementation of tile from the slice. Also it seems backwards to me to compute the slice of the operands, then pass it to the tiling implementation to say produce the tile. It is much more natural to just pass the slice of the iteration space and let the operation implementation handle the slicing/dicing of the operands/results. Adding more data structures, and semantics should all layer on top of this as a "specialization".  In my mind, if thats the core method then that betrays the motivation behind how `TilingInterface` is structured.

Reply:
> 1. `getIterationSpaceTileFromResultTile` - given information of a tile of the result, compute the tile of the iteration space that when used with `getTiledImplementation` generates the required result tile.

Hi, @MaheshRavishankar, this helper method has already been built in [this PR](https://github.com/llvm/llvm-project/pull/93144), would you help to take a look?

Reply:
> Ok, I think this PR has been put through the review wringer a bit too much. I dont think we should use PRs as a way to bike shed ideas. Maybe some clarification of what the interface is set up to do might help restrict the scope of expansions/modifications to first land this, and then think of how to change the interface to handle it. In general I have strong concerns with the direction of changes suggested in this comment thread. Ill answer those later, but first let me write down what each method in the interface does.

Thanks for this write-up! Would you mind putting it where it belongs, that is, in the [interface documentation](https://github.com/llvm/llvm-project/blob/3ce9b86cfd2d88162bc4a551dd7910b8cff3097b/mlir/include/mlir/Interfaces/TilingInterface.td#L19-L20).

Indeed, there has been quite a lot of design discussion in this PR, though most of the initial reviews were enforcing code compatibility with the rest of MLIR codebase as well as correctness of the results. Maybe it should have been preceded by an RFC where such a discussion could have taken place.

> getTiledImplementation is designed to be as less opinionated as possible. It deals with just tiling of an operation. <...> In process it **might take slices of the operation/it might not**, that is left to the implementation of the operation.
> getResultTilePosition is also needed for implementing tiling. <...> Most **tensor tiling algorithms need to stitch back** the original tensor value from the tiled operation.

This asymmetry is baffling. Why does operand coordinate space slicing is delegated to the interface method, but result space recomposition is not? (I know the technical issue is due to `insert_slice`/`parallel_insert_slice` being dependent on the loop type.)

> This method might or might not reuse the implementation in getTiledImplementation. That depends on op semantics. That is why there isnt a default implementation of this method in the current interface definition.

It is unclear to me when it cannot reuse `getTiledImplementation`? Do you have a specific operation in mind?

> I think things get muddled when we try to have a default implementation of the two fusion methods. Default implementation now need to take an opinion on how the tiling of each operation works.

It is muddled even before. Why does the tiling interface should know anything about fusion? It is sufficient to provide the way to compute mappings between slices of the iteration space and slices of the operands / results. 

>  They are just convenience methods that reduce implementation burden on operations that implement the interface.

Why shouldn't we reduce the implementation burden? This is probably my main concern here.

> I have strong concerns with this. getTiledImplementation is necessarily black boxed. I would not want to assume that you can produce an implementation of tile from the slice. <...>

Maybe I'm insufficiently clear above. I am _not_ suggesting that `getTiledImplementation` should take slices of operands _instead_ of the iteration space slice. I am suggesting that it takes them _in addition_ to that. Same as @qedawkins in his comment. This is merely splitting current implementations of the `getTiledImplementation` method into two parts (1) one that produces operand slices (or just returns the operand if it for some reason doesn't need to do anything) and (2) one that produces the computational body of the loop given the result of (1). Could you elaborate on why this would be more opinionated?

> I suggest we first try to land this PR with something that is equivalent of generateOperandTileValue and then come back to maybe having default implementation of the different methods.

Does this solve the issue @qedawkins pointed out? (By triplication of logic in each implementation I suspect.)

Reply:
Just for completeness, I have made a similar comment on the patch that introduced `getOperandTileValue` - https://reviews.llvm.org/D127809#inline-1228120 - but made it non-blocking. I was wrong not to force that discussion then.

Reply:
> It is unclear to me when it cannot reuse getTiledImplementation? Do you have a specific operation in mind?

I'm not sure about the current implementation, but I could see this being the case for `tensor.pad`. Tiling based on the operand must produce the entire padded tensor near the boundaries (and thus can always tile to another `tensor.pad`), but tiling based on the full iteration space of the result could instead produce `scf.if { } else` to avoid dynamically zero-sized tensors. In other words, tiling based on the operand could be a restricted subset of tiling for the whole operation and thus want a different implementation.

> Ok, I think this PR has been put through the review wringer a bit too much. I dont think we should use PRs as a way to bike shed ideas.

Good point and I contributed to this. I'm going to avoid extending the conversation here then and wait for an RFC or another place to discuss. I, and it seems others as well, have work dependent on this change so I would like to see it land in some capacity sooner rather than later, but that should not be a forcing function. Otherwise, the lit tests in this PR look good to me so I'll keep my approval as is for those portions.

Reply:
> I'm not sure about the current implementation, but I could see this being the case for tensor.pad. Tiling based on the operand must produce the entire padded tensor near the boundaries (and thus can always tile to another tensor.pad), but tiling based on the full iteration space of the result could instead produce scf.if { } else to avoid dynamically zero-sized tensors. In other words, tiling based on the operand could be a restricted subset of tiling for the whole operation and thus want a different implementation.

I don't completely follow this. I suppose I don't understand what you mean by "tiling based on the operand must produce the entire padded tensor" here. What this seems to imply is that a tileable operation does not necessarily tile to a smaller copy of itself, so a `tensor.pad` could, e.g., tile to a `linalg.fill` for the result slice that entirely consists of padded values. 

I would argue that it is still possible to achieve this with two separate methods as described above, both of which would emit `scf.if` with conditions based on the position in the iteration space, but that could result in suboptimal IR if the canonicalizer / DCE is not invoked. Still strange to me that we should push the burden on the interface implementation instead of, well, improving the canonicalizer, but at least makes some sense conceptually.

Thanks for the example by the way. If the somebody had provided that in response to my asking about the method being a trivial composition of two other methods here https://github.com/llvm/llvm-project/pull/88712/files#r1604607877, it would have saved a lot of time.

Reply:
> 
> Thanks for this write-up! Would you mind putting it where it belongs, that is, in the [interface documentation](https://github.com/llvm/llvm-project/blob/3ce9b86cfd2d88162bc4a551dd7910b8cff3097b/mlir/include/mlir/Interfaces/TilingInterface.td#L19-L20).

Yes, Ill do that. I realize this is probably useful to document more than it is today.

> 
> Indeed, there has been quite a lot of design discussion in this PR, though most of the initial reviews were enforcing code compatibility with the rest of MLIR codebase as well as correctness of the results. Maybe it should have been preceded by an RFC where such a discussion could have taken place.

Maybe. From my view this was a strict extension to the interface it didn't cross my mind that it needed an RFC. I didn't anticipate that the change would be controversial.


> This asymmetry is baffling. Why does operand coordinate space slicing is delegated to the interface method, but result space recomposition is not? (I know the technical issue is due to `insert_slice`/`parallel_insert_slice` being dependent on the loop type.)

Good question! Comes down to how each loop handles the tiles. While the interface method handles the "generation of a tile of the result, and by extension taking the appropriate slices of the operands (or whatever else it needs to do to generate the tile of the result), how these tile results are put back is actually upto the loop type. It doesnt seem as the right break up to push that to the implementation itself. Then each implementation will have to account for each loop type which breaks the purpose of the interface.

> 
> > This method might or might not reuse the implementation in getTiledImplementation. That depends on op semantics. That is why there isnt a default implementation of this method in the current interface definition.
> 
> It is unclear to me when it cannot reuse `getTiledImplementation`? Do you have a specific operation in mind?

I think Quinn pointed out the `tensor.pad` implementation. Honestly I dont have all of the implementations indexed in my head to think of an example. The goal is to build the `TilingInterface` to expose the minimum amount of information that an operation needs to expose in order to be usable to write a tiling implementation in an op-agnostic way. Its a bit of a path finding in some cases, and making as less assumptions as possible. 


> 
> > I think things get muddled when we try to have a default implementation of the two fusion methods. Default implementation now need to take an opinion on how the tiling of each operation works.
> 
> It is muddled even before. Why does the tiling interface should know anything about fusion? It is sufficient to provide the way to compute mappings between slices of the iteration space and slices of the operands / results.
> 

One of the aims of `TilingInterface` was to "generalize" what Linalg does and make it more widely applicable. We havent made that much progress on this in LLVM itself (more progress has been made in IREE). So the tiling interface was done not only to generalize tiling but also generalize tile + fuse. So the `getTiledImplementation` only deals with tiling. If an op implements more methods then it can be tiled and fused as well. So they are inherently linked. I have also stayed away from saying slice of iteration space -> slice of operands/results is the "requirement" for tiling. Instead, my base assumption is not too rely on any relationship between iteration space and operand/result slices. But it is also true, that if this relationship exists, then a default implementation could cover the use case for a lot of operations. At the outset, I wasnt sure that this is actually the case, but as more of these have been added, it does seem to be so.

> > They are just convenience methods that reduce implementation burden on operations that implement the interface.
> 
> Why shouldn't we reduce the implementation burden? This is probably my main concern here.
> 
> > I have strong concerns with this. getTiledImplementation is necessarily black boxed. I would not want to assume that you can produce an implementation of tile from the slice. <...>
> 
> Maybe I'm insufficiently clear above. I am _not_ suggesting that `getTiledImplementation` should take slices of operands _instead_ of the iteration space slice. I am suggesting that it takes them _in addition_ to that. Same as @qedawkins in his comment. This is merely splitting current implementations of the `getTiledImplementation` method into two parts (1) one that produces operand slices (or just returns the operand if it for some reason doesn't need to do anything) and (2) one that produces the computational body of the loop given the result of (1). Could you elaborate on why this would be more opinionated?
> 

(1) is fundamentally assuming that you can compute the operands slices and that you can use it to create the tiled implemetnation. It is indeed true for Linalg and other operations, but I find it hard to say that is always possible to split this way, and I would say the split would be justified if it were needed by the implementation of tiling that is using the interface. If the tiling implementation does not need to have this split, then there is no reason for the interface to have these methods.

> > I suggest we first try to land this PR with something that is equivalent of generateOperandTileValue and then come back to maybe having default implementation of the different methods.
> 
> Does this solve the issue @qedawkins pointed out? (By triplication of logic in each implementation I suspect.)

Follow up would be to have the default implementation just use the mapping from iteration space <-> data space and reuse `getTiledImplementation`. That will reduce the duplication for cases it is indeed a duplication and preserve the ability of operations to override it when it doesnt match the op semantics.

Reply:
> > 1. `getIterationSpaceTileFromResultTile` - given information of a tile of the result, compute the tile of the iteration space that when used with `getTiledImplementation` generates the required result tile.
> 
> Hi, @MaheshRavishankar, this helper method has already been built in [this PR](https://github.com/llvm/llvm-project/pull/93144), would you help to take a look?

Hi @Yun-Fly , it is on my list to review. I need to look deeper into it.

Reply:
@qedawkins  and @ftynse thanks for the time taken to review the PR. I didnt mean to suggest that it wasnt useful earlier. Was just attempting to step back and say "something here isnt ideal". Actions items now are
1) Land this PR
2) Update `TilingInterface` documentation to capture things said above
3) Follow up offline to get better mind-share of different view points and an RFC for changes if they are indeed needed.